src/flags.zig

```diff
@@ -296,90 +296,135 @@ fn parse_value(comptime T: type, flag: []const u8, value: [:0]const u8) T {
     };
 
     if (V == []const u8 or V == [:0]const u8) return value;
-    if (V == ByteSize) return parse_value_size(flag, value);
+    if (V == ByteSize) return ByteSize.parse(flag, value);
     if (@typeInfo(V) == .Int) return parse_value_int(V, flag, value);
     if (@typeInfo(V) == .Enum) return parse_value_enum(V, flag, value);
     comptime unreachable;
 }
 
-pub const ByteSize = struct { bytes: u64 };
-fn parse_value_size(flag: []const u8, value: []const u8) ByteSize {
-    assert((flag[0] == '-' and flag[1] == '-') or flag[0] == '<');
+pub const ByteUnit = enum(u64) {
+    bytes = 1,
+    kb = 1024,
+    mb = 1024 * 1024,
+    gb = 1024 * 1024 * 1024,
+    tb = 1024 * 1024 * 1024 * 1024,
+};
 
-    const units = .{
-        .{ &[_][]const u8{ "TiB", "tib", "TB", "tb" }, 1024 * 1024 * 1024 * 1024 },
-        .{ &[_][]const u8{ "GiB", "gib", "GB", "gb" }, 1024 * 1024 * 1024 },
-        .{ &[_][]const u8{ "MiB", "mib", "MB", "mb" }, 1024 * 1024 },
-        .{ &[_][]const u8{ "KiB", "kib", "KB", "kb" }, 1024 },
-    };
+pub const ByteSize = struct {
+    value: u64,
+    unit: ByteUnit,
 
-    const unit: struct { suffix: []const u8, scale: u64 } = unit: inline for (units) |unit| {
-        const suffixes = unit[0];
-        const scale = unit[1];
-        for (suffixes) |suffix| {
-            if (std.mem.endsWith(u8, value, suffix)) {
-                break :unit .{ .suffix = suffix, .scale = scale };
+    pub fn default(value: u64) ByteSize {
+        return .{
+            .value = value,
+            .unit = ByteUnit.bytes,
+        };
+    }
+
+    fn parse(flag: []const u8, value: []const u8) ByteSize {
```

Comment:
It might be, but there's a preexisting error handling pattern in this file where every function calls `fatal` directly on errors instead of propagating them; and several parsing functions take this same `flag` parameter as part of that pattern.

If the error handling is not correct here I think it should be addressed for the entire file at once. Can I (should I) file a followup issue about it instead of changing it here?

Reply:
I don't think it is necessary. Just that it feels odd when we have encapsulated “parse” as a type function that should know nothing about the error handling. I’d suggest you to keep the error handling function symmetrical to the other ones, but calling the “try ByteSize.parse” that we decoupled. Or, if you find it easier/clearer, you can revert it to a single function outside BiteSize, it is your call. I think the most important is to be able to precisely store and retrieve units either from constants or from the user input.